node/src/pruner.rs

```diff
@@ -125,78 +149,258 @@ where
                 }
-                let height = header.height().value();
-
-                let cids = self
-                    .store
-                    .get_sampling_metadata(height)
-                    .await?
-                    .map(|m| m.cids)
-                    .unwrap_or_default();
-                for cid in cids {
-                    self.blockstore.remove(&cid).await?;
+            }
+
+            for range in prunable_batch.into_inner() {
+                let from_height = *range.start();
+                // Because of `cancellation_token` we need to track up to which
+                // height we removed.
+                let mut to_height = None;
+
+                for height in range {
+                    if self.cancellation_token.is_cancelled() {
+                        break;
+                    }
+
+                    let cids = self
+                        .store
+                        .get_sampling_metadata(height)
+                        .await?
+                        .map(|m| m.cids)
+                        .unwrap_or_default();
+
+                    for cid in cids {
+                        self.blockstore.remove(&cid).await?;
+                    }
+
+                    self.store.remove_height(height).await?;
+                    to_height = Some(height);
                 }
 
-                self.store.remove_height(height).await?;
-                last_removed = Some(height);
+                if let Some(to_height) = to_height {
+                    self.event_pub.send(NodeEvent::PrunedHeaders {
+                        from_height,
+                        to_height,
+                    });
+                }
             }
+        }
 
-            if last_reported != last_removed {
-                last_reported = last_removed;
-                self.event_pub.send(NodeEvent::PrunedHeaders {
-                    to_height: last_removed.expect("last removed height to be set"),
-                });
-            }
+        debug!("Pruner stopped");
+        Ok(())
+    }
 
-            select! {
-                _ = self.cancellation_token.cancelled() => break,
-                _ = sleep(self.pruning_interval) => ()
+    async fn update_cached_data(
+        &mut self,
+        stored_blocks: &BlockRanges,
+        sampling_cutoff: &Time,
+        pruning_cutoff: &Time,
+    ) -> Result<()> {
+        let update_afer = Duration::from_secs(1).min(self.block_time);
+
+        if self
+            .cache
+            .updated_at
+            .is_some_and(|updated_at| updated_at.elapsed() < update_afer)
+        {
+            // No need to update cached values
+            return Ok(());
+        }
+
+        let after_sampling_window =
+            find_height_after_window(&*self.store, stored_blocks, sampling_cutoff).await?;
+        let after_pruning_window =
+            find_height_after_window(&*self.store, stored_blocks, pruning_cutoff).await?;
```

Comment:
so those do the binary search, which involves potentially a lot of store access. Could we instead cache the next header's time (n + 1 height) and check if it fallen outside of the window to avoid repeating the binary search if we have continuous range of headers? Then we could only update caches with a single query of height + 1 each time.

Reply:
Implemented an algorithm that uses some cached values. Check `find_height_after_window_fast`